### ISSUE TITLE
How about adding type hints?

### ISSUE THREAD

**Issue by GeorgySk** [shapely/shapely#721]
As Shapely currently doesn't have [type hints](https://www.python.org/dev/peps/pep-0484/), results produced by type checkers are far from perfect. Some examples:
```python
from shapely.geometry import Polygon, LineString, Point

area = Polygon().area  # Inferred type: Any. Should be: float
geom_type = LineString().geom_type  # Inferred type: Union[str, List[str]]
# Should be LineString in this case
representative_point = Polygon().representative_point()
# Inferred type: Union[Polygon, Any]. Should be: Point
x = Point(1, 1).x  # Inferred type: Union[object, Tuple[object, object, object], Tuple[object, object]]
# Should be: float
```
and so on.

So, how about adding [stub files](https://mypy.readthedocs.io/en/latest/stubs.html) with the type hints, at least for the most basic functionality from the docs?

---
Shapely version: 1.6.4.post1, installed from conda.

**Comment by sgillies** [comment]
I'm not opposed to adding stub files. However, since I'm not a mypy user (yet) and there is a lot of code to annotate, I am not going to take on this task. I would be happy to review a PR, especially if it had tests of the annotations.

**Comment by snorfalorpagus** [comment]
I would love to see type annotations in Shapely.

**Comment by mwtoews** [comment]
Annotations would be nice to have, perhaps for the 1.8 milestone (?)

One thing to consider planning this feature is whether to put annotations into stub files to maintain Python 2.7 compatibility for future versions of shapely. Or, make the next milestone (e.g. 1.8) release Python 3+ only, and maintain a separate branch for Python 2.7 bugs.

Note that Python 2.7 will not be maintained past 2020 ([about 7 months from now](https://pythonclock.org/)).

**Comment by mwtoews** [comment]
Python 3.4 reached it's [end-of-life in March 2019](https://www.python.org/downloads/release/python-3410/), and #680 addressed that for Travis CI tests, so I think requiring Python 3.5+ for the next milestone (1.8?) makes sense.

But this means we'd need to maintain a 1.7 branch for Python 2.7 users (also <=3.4).

@sgillies is it time to break away from the Python 2.x era?

**Comment by sgillies** [comment]
Let's revisit this after 2.0.

**Comment by sgillies** [comment]
@mwtoews in https://github.com/shapely/shapely/discussions/1316 it looks like we welcome first time contributors to add type hints. I disagree and think that this is work that should be done by core developers who understand the API very well.

**Comment by idanmiara** [comment]
> Let's revisit this after 2.0.

Hi @sgillies,
Would you like to reopen this issue so we can discuss https://github.com/shapely/shapely/pull/1687 at a small scale?

**Comment by idanmiara** [comment]
Hi all,
I wasn't aware of this discussion and made a major effort in https://github.com/shapely/shapely/pull/1687 to close this issue all at once.
I'd make smaller scale PR as @jorisvandenbossche suggested in https://github.com/shapely/shapely/pull/1687#issuecomment-1407043373 so we can discuss the details.

**Comment by sgillies** [comment]
@idanmiara @jorisvandenbossche so here's the deal with typing: because of projects like Pydantic, hints are no longer just hints to be used by mypy etc. The hinted return type of Shapely methods are now part of the public API because people will use them at runtime.

I don't think we know yet whether we want to commit to returning concrete classes or more abstract types. Certainly more discussion is needed.

**Comment by idanmiara** [comment]
> I don't think we know yet whether we want to commit to returning concrete classes or more abstract types. 

Would you please point out a few examples from https://github.com/shapely/shapely/pull/1687 that would 
demonstrate your concern, like what are the concrete classes vs more abstract types that we should consider ?

**Comment by parched** [comment]
> @idanmiara @jorisvandenbossche so here's the deal with typing: because of projects like Pydantic, hints are no longer just hints to be used by mypy etc. The hinted return type of Shapely methods are now part of the public API because people will use them at runtime.

There are a few ways to solve this.
1. State that the types aren't part of the API (simple but relies on people reading the docs)
2. Prefix them all with an underscore or put them in a module prefixed by an underscore (a lot of projects do this for many of their types, including the types for the standard library).
3. Leave the typing to a 3rd party stubs package (easier to get out of sync)

**Comment by idanmiara** [comment]
> 2. Prefix them all with an underscore or put them in a module prefixed by an underscore (a lot of projects do this for many of their types, including the types for the standard library).

like this?
https://github.com/shapely/shapely/pull/1687/files#diff-45b05889fa98905e76c72e27cbc5d4fb92cf61f58b9da4b8778cf8957afe24df

**Comment by parched** [comment]
> like this? https://github.com/shapely/shapely/pull/1687/files#diff-45b05889fa98905e76c72e27cbc5d4fb92cf61f58b9da4b8778cf8957afe24df

Yes, that looks good.

**Comment by idanmiara** [comment]
Hi all,
As per @jorisvandenbossche  suggestion, I've made a smaller PR #1741 (from #1687) that includes only the annotations for `base.py`.
I'd be happy to get your reviews.

**Comment by jorisvandenbossche** [comment]
I personally don't use type annotations regularly (or make use of the features it enables in eg IDEs), but I do understand that other people do and get benefits from it, so I am certainly open to consider how we can improve this for shapely. 
That said, my personal concerns with adding type annotations to the shapely codebase:

- It should not make contributing (much) harder for people that are _not_ familiar with all the complex details of typing in Python. 
  For example, the current `_typing.py` file added in https://github.com/shapely/shapely/pull/1741/files#diff-45b05889fa98905e76c72e27cbc5d4fb92cf61f58b9da4b8778cf8957afe24df feels very complex to me.
- Most of the time, I find that complex type hints make the documentation _less_ readable (eg see [#1741 (comment)](https://github.com/shapely/shapely/pull/1741#discussion_r1090320354)). So if we add type hints, I would personally strongly prefer that we don't include the annotations in the documentation (sphinx' autodoc has an option to not include them in the signature in the generated html documentation; unfortunately in IPython you can't force to hide this)
- How do we ensure the type annotations are actually correct? (and stay correct in the future) I think that ideally, if you have user-facing annotations of the public API, this should also be tested somehow (but I am no expert in typing, so no idea what the typical approach is for this)
- Adding type annotations also brings up the question: how "correct" do we want our type annotations to be? 
  - For example numpy explicitly documents that their type annotations are often stricter than what is allowed at run-time (https://numpy.org/devdocs/reference/typing.html#differences-from-the-runtime-numpy-api)
  - For shapely, we have the similar question for functions that accept array-likes: do we just mark those as accepting `np.ndarray`, or do we actually try to properly define "array-like" in a type alias? (numpy >= 1.20 provides a `np.typing.ArrayLike`)
  - Another question: in most places where we accept or return a Geometry or or array of Geometries, the geometry itself can actually be an actual Geometry or None. Do we want to correctly type this everywhere? This makes hints more verbose (for _development_, that can of course be solved with aliases), but I also wonder how (un)useful this is for actually using the type hints? (because whenever you call a method on the result of a function that could return a Geometry or None, type checkers will (I assume?) complain that `None` doesn't have such method, and you will constantly have to cast `Union[Geometry, None]` to `Geometry` to make the type checking useful).
  - How far do we want to go with distinguishing different geometry types (eg functions returning Point vs generic Geometry)? (and how useful is this for applications? I assume it can be useful to know you have a Point, so that the type checker knows that getting the `.x` attribute is valid?)

**Comment by idanmiara** [comment]
>   For example, the current `_typing.py` file added in https://github.com/shapely/shapely/pull/1741/files#diff-45b05889fa98905e76c72e27cbc5d4fb92cf61f58b9da4b8778cf8957afe24df feels very complex to me.

does it feel very complex to add new types to `_typing.py`? or just to understand it and use it?

> * Most of the time, I find that complex type hints make the documentation _less_ readable (eg see [#1741 (comment)](https://github.com/shapely/shapely/pull/1741#discussion_r1090320354)). So if we add type hints, I would personally strongly prefer that we don't include the annotations in the documentation (sphinx' autodoc has an option to not include them in the signature in the generated html documentation; unfortunately in IPython you can't force to hide this)

Sounds good to me as long as the annotations are embedded in the function signature.

> * How do we ensure the type annotations are actually correct? (and stay correct in the future) I think that ideally, if you have user-facing annotations of the public API, this should also be tested somehow (but I am no expert in typing, so no idea what the typical approach is for this)
> * Adding type annotations also brings up the question: how "correct" do we want our type annotations to be?
>   
>   * For example numpy explicitly documents that their type annotations are often stricter than what is allowed at run-time (https://numpy.org/devdocs/reference/typing.html#differences-from-the-runtime-numpy-api)
>   * For shapely, we have the similar question for functions that accept array-likes: do we just mark those as accepting `np.ndarray`, or do we actually try to properly define "array-like" in a type alias? (numpy >= 1.20 provides a `np.typing.ArrayLike`)

I think that the `Like` suffix notion for the annotations is a good compromise here, i.e. in functions that accept `ArrayLike` (are there other examples besides `ArrayLike` ?)

>  because whenever you call a method on the result of a function that could return a Geometry or None, type checkers will (I assume?) complain that `None` doesn't have such method, and you will constantly have to cast `Union[Geometry, None]` to `Geometry` to make the type checking useful).

It doesn't seem like an issue, at least with PyCharm:

```
def foo(a: int, flag: bool) -> Optional[int]:
    return a if flag else None

def bar(a: int) -> None:
    return None

foo(8, True) + 1  # PyCharm: All good :)
foo(8, False) + 1  # PyCharm: All good (but actually runtime error)
bar(7) + 1  # PyCharm: Expected `int` got `None` instead
```

>   * How far do we want to go with distinguishing different geometry types (eg functions returning Point vs generic Geometry)? (and how useful is this for applications? I assume it can be useful to know you have a Point, so that the type checker knows that getting the `.x` attribute is valid?)

I think that it helps if we annotate the most correct geometry type, 
for example if a function returns always `Point` we should annotate `Point`, and not `Geometry` which seems misleading to me.

**Comment by parched** [comment]
> It doesn't seem like an issue, at least with PyCharm:

Generally, it is an issue. Mypy and Pyright complain about it. For our team, using Pyright, it would negate most of the benefit of the type hints as a lot of `# type: ignore` would be needed. That example can by typed more precisely using overloads, but that won't always be the case in more complex examples.
```python
from typing import overload, Literal, Optional
@overload
def foo(a: int, flag: Literal[True]) -> int:
    ...
@overload
def foo(a: int, flag: Literal[False]) -> None:
    ...
def foo(a: int, flag: bool) -> Optional[int]:
    return a if flag else None
```

As an aside, our team is really excited about the addition of type hints to Shapely, it's the only library left that we use without them. Thanks for working on this @idanmiara.

**Comment by parched** [comment]
> How do we ensure the type annotations are actually correct? (and stay correct in the future) I think that ideally, if you have user-facing annotations of the public API, this should also be tested somehow (but I am no expert in typing, so no idea what the typical approach is for this)

If you're willing to add type hints to the internal functions too, you can run a typechecker over the whole codebase. Otherwise, you can use something like `pyright --verifytypes` that ensures all the public API *has* type hints, but it won't check they're correct AFAIK.

**Comment by idanmiara** [comment]
> That example can by typed more precisely using overloads, but that won't always be the case in more complex examples.

How would you handle the `Optional` issue that @jorisvandenbossche raised ?

**Comment by idanmiara** [comment]
> If you're willing to add type hints to the internal functions too...

I've already annotated (almost?) _all_ the functions and willing to make multiple PRs for them once we have consensus about the basics.

**Comment by kylebarron** [comment]
> * Another question: in most places where we accept or return a Geometry or or array of Geometries, the geometry itself can actually be an actual Geometry or None

I think that most of the time shapely functions are generic over whether it's an array or not. So an array input will always give array output and likewise scalar input will give scalar output. 

You _could_ add overloads for every function but that sounds horribly messy and error prone. I assume it would be much easier with generics. For example:

```py
from typing import TypeVar, reveal_type
import numpy as np
from numpy.typing import NDArray

array = np.zeros(10, dtype=np.int64)
scalar = 1

T = TypeVar('T', int, NDArray[np.int64])

def identity(object: T) -> T:
    return object

reveal_type(identity(scalar))
# tmp.py:14: note: Revealed type is "builtins.int"
reveal_type(identity(array))
# tmp.py:16: note: Revealed type is "numpy.ndarray[Any, numpy.dtype[numpy.signedinteger[numpy._typing._64Bit]]]"
```

This doesn't yet work with actual shapely objects because mypy complains with
```
Skipping analyzing "shapely": module is installed, but missing library stubs or py.typed marker
```
but you should be able to make the `TypeVar` generic over at a minimum
```py
T = TypeVar('T', shapely.geometry.base.BaseGeometry, NDArray[np.object_])
```
Then the static type checker can at least check whether the returned object is a scalar or an array.

At the cost of more complexity, you could even make the `TypeVar` generic over specific types. Something like

```py
T = TypeVar('T', Point, LineString, Polygon, ..., NDArray[np.object_])

def make_valid(geometry: T) -> T:
    ...
```

Then `make_valid(point)` would be statically inferred as returning a point. Not sure if this is possible in the array case

**Comment by idanmiara** [comment]
> You _could_ add overloads for every function but that sounds horribly messy and error prone. I assume it would be much easier with generics.

That's sounds like a good idea! How would you type the following using generics ? 
`def contains(self, other: ScalarOrArrayLike["Geometry"]) -> ScalarOrArray[bool]:`
(input: scalar -> output: scalar, input: array, output: array)

**Comment by jorisvandenbossche** [comment]
> > For example, the current `_typing.py` file added in https://github.com/shapely/shapely/pull/1741/files#diff-45b05889fa98905e76c72e27cbc5d4fb92cf61f58b9da4b8778cf8957afe24df feels very complex to me.
> 
> does it feel very complex to add new types to `_typing.py`? or just to understand it and use it?

Your original version of the `_typing.py` file contained 50+ type alias definitions. So having to understand each of those to know which to use when is certainly a complication IMO. 
(I assume with the other feedback here, that number might go down)

**Comment by idanmiara** [comment]
> > > For example, the current `_typing.py` file added in https://github.com/shapely/shapely/pull/1741/files#diff-45b05889fa98905e76c72e27cbc5d4fb92cf61f58b9da4b8778cf8957afe24df feels very complex to me.
> > 
> > 
> > does it feel very complex to add new types to `_typing.py`? or just to understand it and use it?
> 
> Your original version of the `_typing.py` file contained 50+ type alias definitions. So having to understand each of those to know which to use when is certainly a complication IMO. (I assume with the other feedback here, that number might go down)

True. 
But at least for this #1741 `_typing.py` [has shrunk significantly](https://github.com/shapely/shapely/pull/1741/files#diff-45b05889fa98905e76c72e27cbc5d4fb92cf61f58b9da4b8778cf8957afe24dfR1) 😄

**Comment by parched** [comment]
> > You _could_ add overloads for every function but that sounds horribly messy and error prone. I assume it would be much easier with generics.
> 
> That's sounds like a good idea! How would you type the following using generics ? `def contains(self, other: ScalarOrArrayLike["Geometry"]) -> ScalarOrArray[bool]:` (input: scalar -> output: scalar, input: array, output: array)

It's not currently possible in python typing as far as I know. You'd need a higher-kinded constrained `TypeVar` but that's still an [open issue](https://github.com/python/typing/issues/548).

**Comment by idanmiara** [comment]
@kylebarron @parched @jorisvandenbossche @sgillies 
Thanks again for you inputs!

I've modified https://github.com/shapely/shapely/pull/1741 using the feedback in this discussion.

As the [generics typing solution](https://github.com/shapely/shapely/issues/721#issuecomment-1409394547) @kylebarron suggested [won't fit most functions](https://github.com/shapely/shapely/issues/721#issuecomment-1409722676)
Do you think we should use the `ScalarOrArray` approach I suggested or the [@typing.overload](https://github.com/shapely/shapely/issues/721#issuecomment-1409181165) approach @parched suggested (which might be more correct but would require more efforts and might be harder to followup for new contributions) ?
Personally I think that the `ScalarOrArray` approach would be easier to follow and maintain but I would accept any solutions is agreed by the maintainers.

Does anyone want bring up more suggestions / issues or review https://github.com/shapely/shapely/pull/1741 ?

**Comment by Jeremiah-England** [comment]
Back in late November/early December I created a `shapely-stubs` package for my own use. It covered the entire 1.8 API.

I started stubbing 2.0 API but since I have not transitioned my code to 2.0 yet I haven't needed them enough to finish. Also, the idea with the 2.0 stubs was that they would be good enough to publish on PyPi so I was writing tests as I went and the going was slower.

I didn't announce that I was doing them on this issue because I wanted to get further before opening up a discussion about possibly publishing to PyPi. But this looks like it might be a "speak now or hold your peace" moment for Shapely typing, haha. So here are my thoughts.

I think that a separate `shapely-stubs` package is the best *development path*, and that eventually integrating those stubs into Shapely itself or merging them into typeshed are the best *long-term goals*.

For why a stubs package is a good option for the development path:

- Adoption will be low. In my experience most Python developers just use whatever PyCharm or Pylance gives them without thinking to look for stub packages. Low adoption allows us to make mistakes and overhaul something if we go in a bad direction (e.g. [overloads and unions vs `Any`][0]). We can even labeled it as a WIP in the README. And while adoption will be low, we can still get some real-life usage (and feedback!) from the type annotation enthusiasts during development. That seems valuable.
- We can use complicated type hints without as much concern about maintenance costs in the main codebase. The only concern would be if we want to merge them into Shapely itself at some point instead of merging into typeshed. But I think it is OK to make the type hints somewhat complicated but maximally useful, and then see what people think. We could probably shave them down a little bit or just go the typeshed route if people don't like them.
- We can have a faster release cycle than the main repository, which I think makes sense at the early stages of a project.
- Depending on where the stubs repository lives, and how people feel, a lot of the development could happen without the involvement of main maintainers of Shapely, who might have other things they would rather work on.

I think the main reason to *not* use stubs is that they are not as nice to develop with. As @snorfalorpagus said back in 2019,

> Inline type annotations will be much easier to maintain going forward.

I have a +1 on that comment, and I think inline is a good goal. I would prefer them over the eventual typeshed route. But I think given the relative stability of the Shapely API, and that we are just talking about a _development path_ here and not an end goal, this is not a good enough reason to go with inline instead of stubs right now in light of the bullet points above.

The reason I think that typeshed or inline (or stubs alongside the `.py` files in Shapely itself if we want) are good end goals is adoption. I think there are probably a lot of Shapely users who could benefit from enhanced type annotations who wouldn't think (or just don't know) to look for a stub package.

Assuming for the sake of discussion that we want to go the `shapely-stubs` route, I think we should basically copy what they did for [`pandas-stubs`][2]. Copy the CI, the testing method, etc. That is mostly what I did in [my repository][1] but there is more to do there. And we might want to start with a fresh repository and just copy some of the work I did on `shapely.lib`, which is the only module I finished for 2.0 (the rest of the stuff in `master` there is still for 1.8 so beware).

Logistically, we could put the stubs repository in the Shapely GitHub organization under `shapely/shapely-stubs`. Then at least one of the project members would have to be OK with merging stuff in. That might simplify PyPi key/password handling later on if we for some reason decide to keep the stubs around?

Alternatively, one of the those interested in developing the stubs could host the repository on their profile. I think that only makes sense if none of the Shapely members are interested enough to do the final reviews of the stub code and merge/do releases.

[0]: https://github.com/python/mypy/issues/1693
[1]: https://github.com/Jeremiah-England/shapely-stubs
[2]: https://github.com/pandas-dev/pandas-stubs

**Comment by kylebarron** [comment]
I'm +1 on a shapely-stubs route: minimize disruptions to the development process of Shapely in the short term, while also making data points about the value of the type stubs and feasibility of adding types to shapely proper in the future.

**Comment by idanmiara** [comment]
@Jeremiah-England Thanks for sharing your work and thoughts!
It looks like you have done some nice thorough work for Shapely 1.8 and I'm feeling a bit sad I didn't know about it before spending a few days doing some similar work with Shapely 2, typing the entire library inline in https://github.com/shapely/shapely/pull/1687...

I'm also concerned that like myself, other users won't know they even exist (even if they are published).
Maybe I'm missing something fundamental here (as I haven't used much stubs before), but when I want to understand what a function does I usually click on it in PyCharm and see the implementation along with the annotations, docstring and doctests which gives me a complete picture of what the function actually does. 
How would this work with the stubs? Would I need to look for the function and annotations separately? (this sounds very unnatural to me) 
Or is there a way to view them inlined?

> Inline type annotations will be much easier to maintain going forward.

From a quick glance at [your repo](https://github.com/Jeremiah-England/shapely-stubs) and at [`pandas-stubs`](https://github.com/pandas-dev/pandas-stubs) I would have to agree that it looks difficult to write initially and to maintain going forward as Shapely evolves... Also it looks like there would have to be some non-trivial overhead for maintaining and testing it off the main tree.

What is actually the risk of having some initial version of the inlined typings, get feedback from users and improve or refactor them in the next version based on that feedback? We could state that, the annotations, at least for the initial versions, are experimental so they might get changed frequently (also why we should keep the types in `_typing.py`).
I'm not sure I understand why keeping them in an external repo makes that much easier.

**Comment by Jeremiah-England** [comment]
> I'm feeling a bit sad I didn't know about it before spending a few days doing some similar work with Shapely 2, typing the entire library inline in https://github.com/shapely/shapely/pull/1687...

Yes. Apologies for not commenting earlier. I just checked in on this thread a couple days ago and saw all the activity.

But it is possible we could use your inline type hints as a starting point. There are tools to generate stub files from normal Python files.

**Comment by Jeremiah-England** [comment]
@idanmiara, I only responded to part of your comment on Sunday, but I didn't want to go too long without responding to the rest to here it is.

If I understand your comment, you argue that inline > stubs for three reasons:

- Adoption ("other users won't know they exist...")
- Ease of development and maintenance ("difficult to write initially and to maintain going forward...")
- Documentation and discoverability ("I usually click on it in PyCharm and see the implementation along with the annotations...")

I touched on the first two but completely missed the last one (documentation) in my comment.

On _Adoption_, I think we completely agree. inline >> stubs (unless we publish to typeshed).

On _development/maintenance_, I think we agree a little bit. I think where we maybe disagree here is the size of the issue. I think writing stubs might take some getting used to but it is actually not much more difficult than writing inline. I think that most of the work in developing the types is upfront (stubs or inline) and will be in defining good aliases, generics, overloads, tests and overall _patterns_ for Shapely types. Once all that is in place, any new additions to the Shapely API will be pretty copy/paste. So _maintenance_ is not a very big deal. And I think bullet points 2-4 (if they were numbered) in my original comment are serious enough to even tip the scale for _development_ in favor of type stubs. But I am interested in hearing feedback on them.

On _documentation/discoverability_, I didn't really think about this! Good point. I probably didn't think about it because I usually hover to look at types/docstring. I only go to source if I need to look at the implementation to answer a question.

Overall, I think we both agree that inline > stubs. But in closing you asked a key question:

> What is actually the risk of having some initial version of the inlined typings, get feedback from users and improve or refactor them in the next version based on that feedback?

I'll have to think about this more. Pushing types that we feel like might change to the main Shapely library doesn't feel right to me. But when you ask what the actual risks are I'm not really sure. The main reason I didn't reply before was that I hadn't resolved the dissonance there. I hope to post back with either dissolved feelings or reasons why they are there if someone doesn't resolve it for me in their comment before then.

In summary, I think that

- For _adoption_ inline or typeshed >> shapey-stubs. We agree.
- For _development_ stubs > inline. We agree on some specifics, but disagree overall. And I still need to answer your risk question.
- For _maintenance_ inline > stubs. We agree, but sounds like we disagree about the scale of the issue.

**Comment by ciscorn** [comment]
I'm personally creating [shapely-stubs](https://github.com/ciscorn/shapely-stubs) (for shapely 2.0). I would be happy to transfer it to the community.

**Comment by idanmiara** [comment]
@Jeremiah-England @ciscorn nice work! 
From the fact that all theree of us implemented almost complete type annotations for Shapely from scratch without knowing each other's efforts I assume the following:
* the discoverability issue is not minor. 
* type annotations are important enough for each of us as to make this large effort individually...

**Comment by sgillies** [comment]
@Jeremiah-England thank you for proposing that the project begin with stubs. I'm convinced that it is the most pragmatic approach.

I'm very curious to see how similar or different these three efforts are. If they are all the same, it would tell us that typing Shapely is a solved problem and that after a release cycle with stubs on the side we should move quickly to incorporate them into shapely.

If the three efforts are significantly different, it would tell us that there are many unknowns and matters of opinion for typing Shapely and that we should go more slowly.

**Comment by idanmiara** [comment]
> while PR #1687 creates many new types, even ones for various numpy array shapes.

Please note that I made #1741 as the first step as suggested and also reduced and modified the types from the feedback I got here
I did not modify #1687 accordingly, and it should be seen as a an initial draft for typing the entire library.

**Comment by idanmiara** [comment]
@Jeremiah-England @ciscorn @sgillies @jorisvandenbossche 

* We currently have 3 proposed implementations: 
  * from @ciscorn - https://github.com/ciscorn/shapely-stubs (complete[?] shapely 2 stubs)
  * from @Jeremiah-England - https://github.com/Jeremiah-England/shapely-stubs (complete shapely 1.8 stubs + initial shapely 2 stubs)
  * from myself - #1687 (`base.py` annotations, updated according to this discussion) + #1741 (complete shapely 2 annotations) - as @Jeremiah-England suggested out - There are tools to generate stub files from normal Python files - so we can use this a third source

I would like to propose the following action items for `shapely-stubs`:
* First we need to decide which one is our base implementation - I propose https://github.com/ciscorn/shapely-stubs as @ciscorn states it is complete so it can be the initial commit - what do you think?
* Create `shapely-stubs` repository under `https://github.com/shapely` (or move one of existing stub repos to this new place)
* Invite the shapely repository maintainers as well as @Jeremiah-England @ciscorn, myself to be the maintainers of `shapely-stubs` (Does anyone else want to propose themselves as maintainers?)
* license - Choose either `BSD 3-Clause License` (same as @ciscorn and Shapely) or `MIT` (as @Jeremiah-England chose) - I don't have any strong preference here.
* Authors/Credits - I propose to name @ciscorn @Jeremiah-England and myself.
* Once we have an initial stubs repo we will make PRs on top of the initial chosen implementation with respect to the inputs from this discussion also adding any missing pieces from the other 2 implementations and from other stub projects as inspiration, like:
  * https://github.com/pandas-dev/pandas-stubs
  * https://github.com/numpy/numpy-stubs (merged into numpy)
  * https://github.com/typeddjango/django-stubs
* Publish `shapely-stubs` on PyPi once we have an initial version. In the meantime I've captured https://pypi.org/project/shapely-stubs/
* Once we have a version we are happy with - we will make a PR to promote the stubs as "Experimental" in Shapely's github readme / docs pages - to get more feedback from the community.
* Once we have collected enough feedback and approval from the community we will ask the the Shapely maintainers to merge the stubs into the official Shapely repo as inline type annotations - [maybe there would be a way to do this automatically](https://github.com/python/typing/issues/1242)

**Comment by ciscorn** [comment]
@idanmiara  Thank you for putting the current situation into perspective.

If no one has any major concerns about my [_typing.pyi](https://github.com/ciscorn/shapely-stubs/blob/main/shapely-stubs/_typing.pyi), you can use my repository as a foundation.

I think @idanmiara's [_typing.pyi](https://github.com/shapely/shapely/pull/1687/files#diff-45b05889fa98905e76c72e27cbc5d4fb92cf61f58b9da4b8778cf8957afe24dfR33) is more specific and better. It would be good to import this later.

**Comment by idanmiara** [comment]
Does anyone have any thoughts about my proposal ?
Any other ideas of how to proceed from here ?

**Comment by idanmiara** [comment]
@Jeremiah-England @ciscorn @sgillies @jorisvandenbossche
I see that further comments here, Do we have an agreement here ?
Any thoughts about my suggestion https://github.com/shapely/shapely/issues/721#issuecomment-1426998001 ?

**Comment by idanmiara** [comment]
Hi all, 
Is there anyone who's up for this task ?

**Comment by HaraldTR** [comment]
+1 stubs would be fantastic.

**Comment by JohnLunzer** [comment]
Bump, @idanmiara sounds like you're last one standing? Maybe just merge yours?

**Comment by idanmiara** [comment]
> Bump, @idanmiara sounds like you're last one standing? Maybe just merge yours?

I don't mind rebasing my work and rewrite some ideas as the per the maintainer's comments but I'm afraid there's no concensus for merging type hints into the the Shapely repository.

**Comment by JohnLunzer** [comment]
> > Bump, @idanmiara sounds like you're last one standing? Maybe just merge yours?
> 
> I don't mind rebasing my work and rewrite some ideas as the per the maintainer's comments but I'm afraid there's no concensus for merging type hints into the the Shapely repository.

To be more clear, I'm not seeing the support needed to receive consensus. Shapely is a widely used package and at this point any type hints, regardless of their quality (and it certainly seems you've made an honest effort), are going to be of more value to the community than no type hints.

**Comment by idanmiara** [comment]
> To be more clear, I'm not seeing the support needed to receive consensus. Shapely is a widely used package and at this point any type hints, regardless of their quality (and it certainly seems you've made an honest effort), are going to be of more value to the community than no type hints.

I totally agree but I don't have a vote.
Maybe the maintainers would agree to merge them into a dedicated branch and release an alpha version with them to get wider community feedback and subsequently after a few months decide if they are ready to be merged into master.

**Comment by idanmiara** [comment]
As I see no enthusiasm for actively initiating the stubs project (including from myself)...
And as clearly there is a strong community request for type annotations in shapely as demonstrated in this issue and in related PRs and issues,
I would like to re-propose adding some minimal type annotations to shapely master.
Maybe something like I suggested in https://github.com/shapely/shapely/pull/1741 (any other ideas are welcome).
Then adding more type annotations in small doses after we have some community feedback.

**Comment by hamdanal** [comment]
For those interested in type checking projects using shapely, I added complete shapely stubs to [typeshed](https://github.com/python/typeshed). The stubs are now available on PyPI https://pypi.org/project/types-shapely (installable with `pip install types-shapely`). The stubs will be also shipped with VS Code’s Pylance in the next release after they update their copy of typeshed.

**Comment by HaraldTR** [comment]
> For those interested in type checking projects using shapely, I added complete shapely stubs to [typeshed](https://github.com/python/typeshed). The stubs are now available on PyPI https://pypi.org/project/types-shapely (installable with `pip install types-shapely`). The stubs will be also shipped with VS Code’s Pylance in the next release after they update their copy of typeshed.

Great job. I have waited for this day for a long time :D

**Comment by johnthagen** [comment]
If https://pypi.org/project/types-shapely is to be the "official" way to get type checking for shapely, could we get a reference to this in the top level README/PyPI page so that it easier for users to discover than reading through this thread?
